src/Kestrel.Core/Internal/Http2/Http2Stream.cs

```diff
@@ -64,13 +66,29 @@ protected override void OnReset()
 
         protected override void OnRequestProcessingEnded()
         {
-            var states = ApplyCompletionFlag(StreamCompletionFlags.RequestProcessingEnded);
-
             try
             {
+                // https://tools.ietf.org/html/rfc7540#section-8.1
+                // If the app finished without reading the request body tell the client not to finish sending it.
+                if (!EndStreamReceived && !RstStreamReceived)
+                {
+                    Log.RequestBodyNotEntirelyRead(ConnectionIdFeature, TraceIdentifier);
+
+                    var states = ApplyCompletionFlag(StreamCompletionFlags.Aborted);
```

Comment:
A lot of the Abort code doesn't apply since we know the application has already exited.

Reply:
I know a lot of the logic like cancelling the RequestAborted token is redundant. In fact we would need to be careful to call Reset() before Abort() to avoid triggering any registrations in this case.

Still, I think it's worthwhile to keep our Abort() logic consolidated. I'm worried in the future something important could be added to Abort() that will be skipped if OnRequestProcessingEnded() is called before Abort() leading to some subtle bug.